### ISSUE TITLE
hypot() on intervals

### ISSUE THREAD

**Issue by bocc** [JuliaIntervals/IntervalArithmetic.jl#306]
Apparently some of the magic that hypot() does to avoid overflows (and underflows?) interferes when you use intervals. For example,

hypot(-3..4, 1..4) results in [0..\infty] in Julia 1.1, on Windows.

It is not a problem per se (as the sqrt version works well), I just thought you might find it interesting. Also, the package is excellent, thank you!

**Comment by dpsanders** [comment]
Confirmed:
```
julia> hypot(-3..4, 1..4)
[0, ∞]
```

This is pretty bad... Would you like to submit a PR with a trivial implementation of `hypot` for `Interval`s?

**Comment by bocc** [comment]
Yes, I will try to do so, hopefully today, tomorrow the latest.

**Comment by bocc** [comment]
So, I've created a PR, but didn't add a test case. I would suggest something along the lines of:
~~~
√2..5 ⊆ hypot(-3..(-1),1..4)
hypot(-3..(-1),1..4) ⊆ (√2-eps())..(5+eps())
~~~
as this returns `[1.05409, 12.3694]` with the default `hypot` implementation. Also, I wanted to mention that I didn't know how to approach the `Complex{Interval{T}}` case. In the code below:
~~~
using ValidatedNumerics

a = (@interval 0 2) + (@interval 1 5)*im
b = (@interval 1 3) + (@interval 2 4)*im

# a real interval
hypot(a,b)

# a complex interval
sqrt(a^2 + b^2)

# with and without rotation
a^2 + b^2
(sqrt(a^2 + b^2))^2
~~~
`hypot` (from `Base`) returns a Real interval, while `sqrt` returns a Complex one. Also, the two consecutive scaling and rotation in the last part broadens the resulting interval a bit.

**Comment by lbenet** [comment]
Thanks for the contribution; for reference, it is #308 

I do not know exactly where the tests should be coded. Yet, I'd like to comment in your observation that `sqrt` and `hypot` behave different for complex numbers. I think the reason is relaated to `^2` (that you use in `sqrt` above) instead of `abs2`; if you redo your example as `sqrt(abs2(a)+abs2(b))` you get the same as `hypot(a,b)`.

**Comment by gwater** [comment]
Note that the issue comes from a binary comparison (`ax < ay`) in `Base.hypot()` which also explains why some Intervals work fine:
```julia
julia> a = Interval(0.2, 0.5);
julia> b = Interval(1, 4);
julia> @show Base.hypot(a, b);
Base.hypot(a, b) = [1.00124, 4.47214]
julia> @show sqrt(a^2 + b^2);
sqrt(a ^ 2 + b ^ 2) = [1.0198, 4.03113]
```
Note furthermore that Base.hypot actually gives a more precise result here. It might be worth checking for cases where `Base.hypot()` is safe to use and returning the more precise result.

**Comment by dpsanders** [comment]
It seems like it should be possible to use Base's hypot to calculate upper and lower bounds for hypot of an Interval.

**Comment by lbenet** [comment]
Currently, `Base.Math.hypot` is being used:
```julia
julia> hypot(-3..4, 1..4)
[0, ∞]

julia> @which hypot(-3..4, 1..4)
hypot(x::T, y::T) where T<:Number in Base.Math at math.jl:541
```

**Comment by krish8484** [comment]
Fixed in #390
